UnityAds/UnityFullscreenAdAdapter.m

```diff
@@ -0,0 +1,168 @@
+                                                  description:[NSString stringWithFormat:@"Unity Ads failed to load a video ad for %@, with error message: %@", placementId, message]];
+
+    MPLogAdEvent([MPLogEvent adLoadFailedForAdapter:NSStringFromClass(self.class) error:loadError], placementId);
+    [self.delegate fullscreenAdAdapter:self didFailToLoadAdWithError:loadError];
+}
+
+#pragma mark - UnityAdsShowDelegate Methods
+
+- (void)unityAdsShowStart:(nonnull NSString *)placementId {
+  MPLogAdEvent([MPLogEvent adWillAppearForAdapter:NSStringFromClass(self.class)], placementId);
+  [self.delegate fullscreenAdAdapterAdWillPresent:self];
+
+  [self.delegate fullscreenAdAdapterAdDidPresent:self];
+  [self.delegate fullscreenAdAdapterDidTrackImpression:self];
+  MPLogAdEvent([MPLogEvent adShowSuccessForAdapter:NSStringFromClass(self.class)], placementId);
+  MPLogAdEvent([MPLogEvent adDidAppearForAdapter:NSStringFromClass(self.class)], placementId);
+}
+
+- (void)unityAdsShowClick:(NSString *)placementId {
+  MPLogAdEvent([MPLogEvent adTappedForAdapter:NSStringFromClass(self.class)], placementId);
+  [self.delegate fullscreenAdAdapterDidReceiveTap:self];
+  [self.delegate fullscreenAdAdapterDidTrackClick:self];
+  MPLogAdEvent([MPLogEvent adWillLeaveApplicationForAdapter:NSStringFromClass(self.class)], placementId);
+  [self.delegate fullscreenAdAdapterWillLeaveApplication:self];
+}
+
+- (void)unityAdsShowComplete:(NSString *)placementId withFinishState:(UnityAdsShowCompletionState)state {
+  MPLogAdEvent([MPLogEvent adWillDisappearForAdapter:NSStringFromClass(self.class)], placementId);
+  MPLogAdEvent([MPLogEvent adDidDisappearForAdapter:NSStringFromClass(self.class)], placementId);
+  [self.delegate fullscreenAdAdapterAdWillDismiss:self];
+  [self.delegate fullscreenAdAdapterAdWillDisappear:self];
+  [self.delegate fullscreenAdAdapterAdDidDisappear:self];
+  [self.delegate fullscreenAdAdapterAdDidDismiss:self];
+}
+
+- (void)unityAdsShowFailed:(NSString *)placementId withError:(UnityAdsShowError)error withMessage:(NSString *)message {
+    if (error == kUnityShowErrorNotReady) {
+        // If we no longer have an ad available, report back up to the application that this ad expired.
+        // We receive this message only when this ad has reported an ad has loaded and another ad unit
+        // has played a video for the same ad network.
+        NSError *showError =[NSError errorWithAdAdapterErrorCode:MPAdAdapterErrorCodeUnknown
+                                                     description:[NSString stringWithFormat:@"Unity Ads video ad has expired with error message: %@", message]];
+        MPLogAdEvent([MPLogEvent adShowFailedForAdapter:NSStringFromClass(self.class) error:showError], placementId);
+        [self.delegate fullscreenAdAdapterDidExpire:self];
+        return;
+    }
+
+    NSError *showError = [NSError errorWithAdAdapterErrorCode:MPAdAdapterErrorCodeUnknown
```

Comment:
If I remember correctly the idea behind ErrorFactory in Richard's implementation was, so we could unify the message format, for example prepend all the messages with `[Unity Ads]`